jdbc-perf-logger-driver/src/test/java/ch/sla/jdbcperflogger/driver/WrappingDriverUnloadTest.java

```diff
@@ -15,65 +15,81 @@
+
         try {
-            final int clientPort = DriverConfig.INSTANCE.getClientAddresses().get(0).getPort();
-            final ServerSocket logReceiver = new ServerSocket(clientPort);
-            final Socket accept = logReceiver.accept();
-            final int remotePort = accept.getPort();
-            accept.close();
+            await().atMost(WAIT_TIME, TimeUnit.SECONDS).until(portInUse(serverPort));
+
+            final Socket client = logReceiver.accept();
+            final int remotePort = client.getPort();
+
+            await().atMost(WAIT_TIME, TimeUnit.SECONDS).until(portInUse(clientPort));
+            await().atMost(WAIT_TIME, TimeUnit.SECONDS).until(portInUse(remotePort));
+
+            client.close();
             logReceiver.close();
 
             WrappingDriver.unload();
-            assertPortIsAvailable(clientPort);
-            assertPortIsAvailable(remotePort);
-            final Integer serverPort = DriverConfig.INSTANCE.getServerPort();
-            if (null != serverPort) {
-                assertPortIsAvailable(serverPort);
-            }
+            await().atMost(WAIT_TIME, TimeUnit.SECONDS).until(portNotInUse(clientPort));
+            await().atMost(WAIT_TIME, TimeUnit.SECONDS).until(portNotInUse(serverPort));
+            await().atMost(WAIT_TIME, TimeUnit.SECONDS).until(portNotInUse(remotePort));
         } finally {
             WrappingDriver.load();
+            await().atMost(WAIT_TIME, TimeUnit.SECONDS).until(portInUse(serverPort));
         }
     }
 
-    protected void assertPortIsAvailable(final Integer port) {
-        ServerSocket s = null;
-        boolean portAvailable = false;
-        int retries = 0;
-        while (!portAvailable && retries < 10) {
-            try {
-                s = new ServerSocket(port);
-                portAvailable = true;
-            } catch (final IOException e) {
-                portAvailable = false;
-                try {
-                    Thread.sleep(500);
-                } catch (final InterruptedException e1) {
-                }
-                retries++;
-            } finally {
-                if (s != null) {
-                    try {
-                        s.close();
-                    } catch (final IOException e) {
-                    }
-                }
+    protected Callable<Boolean> portInUse(final int port) {
+        return new Callable<Boolean>() {
+
+            @Override
+            public @NonNull Boolean call() throws Exception {
+                return !canOpenPort(port);
             }
+        };
+    }
+
+    protected Callable<Boolean> portNotInUse(final int port) {
+        return new Callable<Boolean>() {
+
+            @Override
+            public @NonNull Boolean call() throws Exception {
+                return canOpenPort(port);
+            }
+        };
+    }
+
+    protected boolean canOpenPort(final int port) {
+        try (
+                ServerSocket serverSocket = new ServerSocket(port)) {
```

Comment:
if the call succeeds, shouldn't you close it in a finally statement? or even better in a try-with-resource since ServerSocket implements AutoCloseable

Reply:
try-with-resource *is* already being used, right?

Reply:
no, try-with-resource is the java 7 feature that looks like this:
try(ServerSocket serverSocket = new ServerSocket(port)) {
}

which always closes the serverSocket after the try

Reply:
sorry, my mistake: because of the line break I took the parentheses for a curly brace...